### DISCUSSION TITLE
zh/development/dev-environments

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#150] [Announcements]
# zh/development/dev-environments

一份非官方的新手指南

https://nixos-and-flakes.thiscute.world/zh/development/dev-environments

**Comment by hanjianwei** [comment]
最近尝试用NixOS跑基于Python的[labelme](https://github.com/labelmeai/labelme)，感觉相当痛苦。当然，Python混乱的打包系统难辞其咎，但是NixOS让这个过程更难了，特别是当Python需要调用系统其他部分时更是如此。Docker之类当然挺好，但是不适合有GUI的应用。

**Reply by ryan4yin** [reply]
Python 确实是个大坑... GUI 的话不知道直接开个 FHS 环境行不行。

**Reply by nowhereman1999** [reply]
docker跑gui也还好把，创建容器时透传一下主机相关的环境变量就可以吧（

**Comment by hanjianwei** [comment]
> docker跑gui也还好把，创建容器时透传一下主机相关的环境变量就可以吧（

话虽如此，不过感觉还是有点麻烦的，特别是临时用到一个东西，发现nix里面没有现成的方案，再临时去研究确实浪费时间。而开一个docker，其实本质上就是开一个沙盒让你放弃nix的纯净环境去随便折腾。用了一段NixOS之后，我又切换到Arch了，然后用Nix（不是NixOS）来管理部分环境。

Nix的想法是好的，可惜现实世界是混乱的，Arch（or Ubuntu/Debian) + Nix相当于在混乱中创造一个纯净环境，而NixOS也应该搞个方便地创造混乱子环境的方法😊

**Reply by nowhereman1999** [reply]
确实，哥们这两天在wsl2上装了下NixOS准备试试，主要想将自己的开发环境打包管理，不用没太机子都重新配置一遍环境了（

**Reply by CS-liujf** [reply]
我用的ubuntu，然后用nix，发现全局安装python包很麻烦，还没想到办法来解决

**Comment by CS-liujf** [comment]
我今天用nix+home-manager(**我的系统是ubuntu**)，发现python没法直接用pip，然而有些包在nixpkg里没有，我就不知道怎么办了。我先说一下自己的需求，原先我用miniconda，它自带全局的虚拟环境base。在这个环境下，我执行pip install都是装在此虚拟环境下，这给我带来了巨大的好处。当我使用vscode的jupyter notebook时，第一次使用它会让我选择解释器，选择好解释器后它会告诉我需要安装ipython等python包，我只需要点install，它全部搞定了（我推测它直接执行pip安装）。以后我再用vscode打开其它路径下的jupyter notebook，只要选择相同的解释器(或虚拟环境)，它都能正常使用。我的问题是，现在我使用nix+home-manager，该如何做到同样的效果，我想全局弄一个base虚拟环境，使得我可以简单执行pip

**Reply by ryan4yin** [reply]
Nix 无法做到同样的效果，你这个需求我还是建议换回 conda 虚拟环境。

**Reply by ryan4yin** [reply]
如果你希望拥有一定的可复现能力，conda 环境也能通过 `conda env export` 导出一个依赖列表（类似 `pip freeze`），备份下这个文件就行。

**Reply by CS-liujf** [reply]
谢谢。但是我现在用home-manager，它管理者zshrc文件，导致没法安装conda，因为conda会往zshrc里添加东西。
是否有方法，使得当前全局能使用pip？

**Reply by ryan4yin** [reply]
你可以直接通过 nix 安装 conda，直接在你的 `home.packages` 里加个 `conda` 就行，然后在任意终端中执行 `conda-shell` 即可进入 conda 环境。

**Reply by CS-liujf** [reply]
我在nixpkg里搜anaconda或者miniconda都没有

**Reply by CS-liujf** [reply]
此外，我用的是ubuntu，然后用home-manager来管理一些配置文件

**Reply by ryan4yin** [reply]
> 你可以直接通过 nix 安装 conda，直接在你的 `home.packages` 里加个 `conda` 就行，然后在任意终端中执行 `conda-shell` 即可进入 conda 环境。

我想我这个描述足够清楚了。

**Reply by CS-liujf** [reply]
我再详细具体说一下，我原先用的bash和miniconda，miniconda往我bash里加了一些它初始化代码。之后，我开始用home-manager，并凭借它安装了zsh，现在转为了zsh，原先的conda没法直接用了。因此我还有个问题想请教，请问是否有方法继续使用原先的conda？

**Reply by CS-liujf** [reply]
非常感谢，我再nixpkg里搜索conda找到了，之前一直只搜过anaconda或miniconda

**Reply by ryan4yin** [reply]
conda 环境的默认路径应该是固定的，你按照我上面说的方式进入 conda 环境，应该有可能直接继续用上你原先的 conda 环境。但我没这么用过，你自己测测吧。

**Reply by ryan4yin** [reply]
如果你想避免使用 nix 安装 conda，当然也可以，这需要你自己将 conda 生成的 zsh 初始化脚本插入到你的 zshrc 中。
举个例子，我在 macOS 上是这么处理的：

https://github.com/ryan4yin/nix-config/blob/main/home/darwin/shell.nix#L10-L17

我这么干的主要原因是 Nixpkgs 的 Conda 不支持 macOS 系统。

**Reply by CS-liujf** [reply]
非常感谢！！！我正是需要一个具体的例子。我原本设想的方法，大致也是往extraEnv后者initExtra里加代码。
再次感谢

**Comment by loskyertt** [comment]
最近在用 WSL2 的 NixOS，想通过 flake 创建一个 Cpp 的开发环境，目前最大的问题就是通过 vscode 连接到 NixOS 后，clangd 这类 lsp 工具都没法使用，甚至连标准库的找不到。后来通过 direnv 解决后，又死活识别不到 OpenMP 库，在 CMakeLists.txt 中配置了都不管用，但是构建项目和编译的时候都能通过......一直解决不了，感觉这种环境隔离配置还是挺麻烦的

**Comment by ynty** [comment]
uv/pixi 这类新生代的 Python 项目/依赖管理工具提供了 lockfile 能力，能一定程度上保障可复现性。

不过问题是：
1. 这两个工具在 Nix 上能否正常工作？
2. 使用这两个工具在 Nix 世界是否属于 best practice？

**Reply by kyv001** [reply]
我自己的少量经验是，uv确实*可以*用，但一些pypi包可能会隐性依赖FHS/系统库，导致工作不正常。我在迁移到NixOS发现pygame窗口不显示了，我觉得可能是对sdl2和FHS的依赖导致的，比如期望在/usr/lib64找到sdl.so.2什么的

**Reply by kyv001** [reply]
pixi可以选择用conda作为后端，而conda也会管理二进制依赖库，可能会更好一些
